### ISSUE TITLE
Follow XDG basedir spec on Linux

### ISSUE THREAD

**Issue by dreamer** [dosbox-staging/dosbox-staging#206]
It's widely recognized and followed spec for desktop Linux software: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

TL;DR; configuration file should be generated by default in `~/.config/dosbox/<name>.conf` instead of `~/.dosbox/<name>.conf`, but XDG env variables need to be followed as well. Old location should still work as it does if `~/.dosbox/` dir is present on user's system.

**Comment by es20490446e** [comment]
Exactly.

**Comment by es20490446e** [comment]
The old config files could be moved to the new directory, and be only read from there.

Keeping that moving code only between consecutive releases, then assume the new location only.

**Comment by dreamer** [comment]
This would be a bit contentious issue (users might have their own scripts, links, etc) - it's not about distros, really - all serious desktop environments adopted XDG years ago. Also, DOSBox upstream will likely never implement this (they refused to accept patches with much less intrusive changes already), but it does not affect us (we have a separate .conf file already).

I think I will handle it this way:

```
new_location <- configuration directory as defined by XDG spec (it's important to handle it exactly as spec says due to Flatpaks and Snaps creating app-specific directory hierarchies)
old_location <- ~/.dosbox/

if $new_location/<conf-name>.conf file exists:
    read settings from this file
else if $old_location/<conf-name>.conf file exists:
    read settings from this file
else
    create new settings file in $new_location/
```
This will give users clear update path (with possibility of migrating manually if they wish so), without breaking anything.

All documentation and help strings need to be updated to point to a new location only, plus a warning to be displayed on stderr on start. As users will install dosbox-staging for the first time the modern location will be picked by default, but our testers, devs, and users who prefer old location won't need to switch right away (and they can even keep using old location if they really want to).

There's also a bit problematic issue of new location for screenshots, recordings, GUS patches and user-supplied OpenGL shaders - some of them will need to move to XDG_DATA_HOME by default (I will need to work on details though).

**Comment by dreamer** [comment]
A little bit related to #73

**Comment by es20490446e** [comment]
Creativity is improvement,
improvement is change.

**Comment by i30817** [comment]
You should try to upstream this anyway (so code it in C not C++). Even if upstream doesn't merge it right away, just having the patch will mean the distro maintainer is more likely to merge it to the package (debian does this often). And if it is in the package patches, it usually ends up upstream sooner or later.

**Comment by dreamer** [comment]
I was informed by DOSBox maintainer, that no new features will be accepted upstream via patches - only bug fixes improving game compatibility, so this change will not qualify for sending upstream.

**Comment by i30817** [comment]
Just talk to the person doing the debian package. They'll do it and it ends up in the users hands that way.

**Comment by dreamer** [comment]
I would much rather be interested in Debian packaging dosbox-staging in the repos. At this point dosbox-staging is **643** patches ahead of upstream SVN - I see no reason why I should treat this change any different than others.

**Comment by i30817** [comment]
Being a debian (or rpm for that matter) packager is real work that people prefer to avoid if there is already something. Sadly i actually hate flatpak and the other projects of that kind so i won't recommend that as a alternative you control. 

Tossing a few simple patches to the debian package maintainer that debian would like is a good way to pressure upstream to actually do something or at least getting that functionality to the relevant users without getting into the probably tiresome politics of joining the debian project and running their linters or be delisted, so that's why i suggested it.

btw if you want some comedy the wall of shame in my ubuntu 19.10 install (programs and frameworks that place hidden things on the HOME dir):

.ant (ancient)
.audacity-data (lol)
.bazaar (godamnit canonical)
.cargo (GODAMNIT MOZILLA)
.dosbox
.exult (another ancient project with little movement)
.gnupgp (i suppose they're afraid a bunch of stuff breaks if they move this)
.java (ditto)
.launchpadlib (wtf is this... apparently switching my repositories from bzr to git in launchpad wasn't without consequences? Delete and hope for the best).
.lyrics (really?)
.mono (of course)
.mozilla (10 years after the spec comes out...)
.ncmpcpp (frontend to the music daemon which is probably dead software that i only run because it's the best way to run mod files i found because everything else is fucking broken...)
.ssh (same deal as gnupgp)
.steam (another barbarian program)
.subversion (from aeons past)
.t-engine (the TOME4 game dir, another case of a game engine not being a good citizen).


So, this has actually became much worse - or i installed more things that are not good citizens. A few releases ago i only had .mozilla, .dosbox and .exult as the main non-framework/crypto dirs there.

**Comment by es20490446e** [comment]
What was the reasoning for not accepting new features?

Cause I'm starting to suspect that the original DosBox project couldn't be the best option long term.

And I think that patching at the package level is undesirable nearly all the time. The package code tends to rot. Better to fork!

**Comment by es20490446e** [comment]
For instance I'm thinking about features that would be very interesting for DosBox to have. Like auto-mounting disk images withing the working folder.

**Comment by dreamer** [comment]
> What was the reasoning for not accepting new features?

"It needs to work everywhere". By "everywhere", they also mean Windows 9x, Windows NT 3.5x or newer, Windows 2000, OS/2, Mac OS X 10.4, and other historic platforms (it's unfeasible to verify builds on all these platforms before sending patches, and SDL2 does not support them anyway).

"It can't change code too much" - because DOSBox is a hobby project, maintainers don't have time for "professional" code review, so no refactorizations are allowed. Changes from "outside" of "DOSBox team" are basically never accepted; some changes we landed in our master branch were waiting for inclusion for years (e.g. 6 years for SDL2 port).

There were also non-technical reasons communicated only in private.

In the context of **this** feature - they already expressed [an opinion](https://www.vogons.org/viewtopic.php?p=833221#p833221):

> But as for the location of the folder, ~/.config is not the correct one, as it doesn't have just the configs but it's the location of a lot of other things.

which is blatantly false, and indicates, they have no clue about developing Linux desktop applications nor what XDG spec is.

If you want to follow our attempts to upstream changes, then: [0](https://www.vogons.org/viewtopic.php?p=695839#p695839), [1](https://www.vogons.org/viewtopic.php?p=790065#p790065), [2](https://sourceforge.net/p/dosbox/patches/283/) (2 patches accepted via this thread were basically rewritten by maintainer, even when it was not necessary), [3](https://sourceforge.net/p/dosbox/patches/284/), (after that maintainers communicated with us only in private).

> Cause I'm starting to suspect that the original DosBox project couldn't be the best option long term.

Yup :( I started posting suggestions for improvements in May 2019 and moved on with this project in September 2019. Gave up on cooperation with upstream in late February 2020. When we will be ready for 0.75.0 release I will try to send few fixes to blatant bugs upstream (e.g. buffer overflows), but I doubt even those will be accepted.

> And I think that patching at the package level is undesirable nearly all the time. The package code tends to rot. Better to fork! (or contribute to a fork)

Maintaining a fork is an uphill battle, especially when original project is old and entrenched amongst users, but upstream refuses to cooperate, so we have no other choice :( Before starting this repo, I investigated a number of other forks ([list](https://github.com/dreamer/dosbox-staging/wiki/DOSBox-forks)), but none of them checked all of my requirements. DOSBox-X came close, but that project is focused on Demoscene (not games), and has plans to ditch SDL altogether (while I want to keep using SDL2 to assure support for future APIs, game controllers, etc).

In the context of this discussion (sending patches directly to Debian, bypassing DOSBox project): sending patches directly downstream of our upstream would be really unproductive (but hey - this is Free software, any user can volunteer their time to doing it, we have nothing against it!).

> For instance I'm thinking about features that would be very interesting for DosBox to have. Like auto-mounting disk images withing the working folder.

DOSBox has many obscure and undocumented features - I think good way to start would be by creating a new issue (labelled as "question" or "enhancement" for feature request) so we can discuss it (maybe the thing you want to accomplish is already possible).

Anyway - for this repo, all new contributors are welcome, so going straight to PR is also an option ;)

**Comment by es20490446e** [comment]
I think that for the moment we should better talk about it with them:
https://www.vogons.org/viewtopic.php?f=32&t=72517&p=833939#p833939

**Comment by sharkwouter** [comment]
I'm currently working on this. If possible, you can assign it to me

**Comment by dreamer** [comment]
@sharkwouter after you responded in thread GitHub allowed me to assing the issue to you :) Thanks for working on this!

**Comment by DominusExult** [comment]
Before you spread more lies, I am NOT a DOSBox developer, thus my opinion is not the DOSBox developer team's opinion.

And yes, talking with the developers is the way to go. You are mixing up issues. The config location on Linux distros has nothing to do with maintaining compatibility with other OS. The refused patches were a whole different thing.

**Comment by es20490446e** [comment]
@DominusExult We can talk calmly about this, or we can conclude the conversation.

What's your choice?

**Comment by es20490446e** [comment]
@dreamer https://youtu.be/5rTOf6FjY3s

**Comment by DominusExult** [comment]
Huh? I've just cleared up certain lies and pointed out a dishonest representation of "upstream" refusing patches.
In the case of this issue, it had been filed in the dosbox bug tracker since 2018 and it wasn't dismissed as "won't fix".

**Comment by es20490446e** [comment]
@DominusExult Sorry, I never deal with abrasive people.

@dreamer Second attempt deleted again from the forum, which suggest me that this is the standard way of dealing of DosBox developers. I will think about forking.

**Comment by es20490446e** [comment]
What about the name "bitbox"?

**Comment by dreamer** [comment]
@es20490446e You are at similar state of mind I was a few months ago - note, that dosbox-staging repo is already a soft fork, so feel welcome to contribute (I really need to work on finishing contributing guide for this project, it's long overdue). If you want to create your own Git-based fork of DOSBox SVN from scratch, take a look at script `scripts/import-from-svn/import-from-svn.sh` - it should save you some trouble :) Results of running this script are available on branches named `svn/*` in this repo. Besides me, this repo has 2 co-maintainers already (and we received a fair number of community contributions as well in these past few weeks), and I am ok with giving more people push access here if it'll make the work go smoother (we all do code review via pull-requests to verify our work before merging it to master anyway, even co-maintainers with push access).

@DominusExult yes, you don't have SVN write access. But you are one of the very few people, who can express an opinion about what should be accepted in DOSBox codebase and be actually listened to. And when you do express an objection, the code won't be merged. Call it a custodian then, if you don't want to call yourself a maintainer.

And contrary to your (repeated) assertions, I DID talk to maintainer with write access. It would be much easier if all our communication was public instead of basically forcing it via PMs and private IRC channel. I tried to keep everything public in Sourceforge patch threads and Vogons thread, but well - that resulted in no progress at all.

The facts are: this issue is an example of yet another topic, where (perfectly valid) user concerns were dismissed and ridiculed by Vogons staff, and bug report was ignored for years. You did scramble to respond in the (now removed) thread only after the word "staging" was mentioned - otherwise, you wouldn't bother. And I was banned from the forum for communication, that happened *outside* of Vogons.

I find the behaviour of Vogons staff to be really disingenuous - patches authored by in-crowd are "always good", patches authored by anyone else are "always bad" - development is either "hobby project" and don't need to be perfect (when authored by in-crowd) or "must be perfect in every aspect" (when authored by anyone else). Bugs are either critical (when it's a pet peeve of in-crowd) or irrelevant (when the issue is raised by anyone else). Development is open (officially), yet closed (in reality). There are no coding guidelines of any kind, yet everyone needs to apply one or the patch will be rejected. Forks are good when it's an easy way to dismiss user's concerns, yet bad when someone starts to work on one. Code needs to work on Windows 9x when it's a way to dismiss the patch, yet it doesn't when in-crowd wants to land a quick patch. It needs to work on OS/2 for sure, but it does not need to be buildable on Windows 10.

In contrast, actual Vogons users (not staff + tiny "elite" clique) are quite friendly and helpful.

**@** **everyone**
From now on, please keep responses in this thread on topic of XDG-compliance; if you want to continue one of several off-topic threads started in here, then create a new issue and label it "question" or "discussion", please.

**Comment by krcroft** [comment]
I was present for and personally attest to those private PM and IRC discussions taking place between Qbix, @dreamer, and myself. Discussions were constructive, honest, and thorough. The revision made by @dreamer in the first-post of the dosbox-staging thread accurately reflects all actionable outcomes from those discussions.

**Comment by es20490446e** [comment]
@krcroft What do you mean? 😯?

**Comment by krcroft** [comment]
The XDG Base Directory Specification came out in 2003 and steadily gained adoption. Eight years after its release, XDG was proposed for DOSBox in 2011, but the suggestion was slighted and dismissed: https://www.vogons.org/viewtopic.php?t=29599  Now another nine years on XDG, now the defacto standard across linux distributions, was proposed again without fanfare (given DOSBox-SVN and dosbox-staging are still not compliant; so the idea applies across the board), but this lead to accounts being banned and the thread being blocked in a power-play of profound contempt and fragility.

@dreamer 's smooth migration proposal for XDG to compliance gets my vote. Package maintainers can pickup dosbox-staging knowing that new deployments will be XDG-compliant from the get-go, while those running prior manual builds of (any) DOSBox can continue to use their `~/.dosbox/<config>`. 

DOSBox already prints where its reading it's config from, and for most linux users it's simply a "double-click-and-paste" to quickly edit.  The concern of longer paths isn't an issue when used this way. Likewise, if we write output to `~/.local/dosbox/`, we can similarly print that path when dosbox starts up.

**Comment by DominusExult** [comment]
@es20490446e: I'm sorry, Vogons can be a harsh place. Moderators don't like posts that don't really state what they want or only post a youtube video of what they want. I'm sure it's not meant like this but a discussion in a forum really doesn't work when one side only posts video instead of writing.

@dreamer you have a really wrong impression of my role in DOSBox development. I'm a moderator on Vogons with a strong opinion but I'm not a gatekeeper of DOSBox development. They don't need my input on that or always listen to me (e.g. otherwise the SDL2 patch and @krcroft's audio library replacements would be in SVN since a long time ago).

Back to the so called facts, no the topic was closed because the OP began posting video instead of writing.
And no I did not "scramble" in there because staging was mentioned. That's pure paranoia.
And the in vs the out crowd... I can't quite put my finger on it but it's likely the way you approach a project with patches and how you react to critique on those. I know being social is hard but it gets you further than being antagonistic all the time.
My reaction to your patches and staging was heavily influenced by my experience with people who approached projects I'm in like you did over the last twenty years.

And really, bad mouthing upstream and vogons staff every (2nd) chance you get, doesn't help. We've tried to make it work on Vogons, we had strong discussions but we were not badmouthing staging or @dreamer but you have really made it difficult when one reads your posts off site. And yes, off site is still something to consider. Would you let someone regularly in your house while he disrespects you outside of the house?

**Comment by sharkwouter** [comment]
People, please. When getting this issue assigned to me it took me long enough to find the requirements for a solution in these comments. Could we please use a different issue for discussion about politics?

**Comment by dreamer** [comment]
The next off-topic response I will move to a new thread myself.

@sharkwouter to make it clearer for future readers; we are aiming at implementation, that functionally works like this:

    new_location <- configuration directory as defined by XDG spec (~/.config/dosbox/ by default)
    old_location <- ~/.dosbox/

    if $new_location/<conf-name>.conf file exists:
        read settings from this file
    else if $old_location/<conf-name>.conf file exists:
        read settings from this file
        log deprecation warning
    else
        create all directories on path to $new_location with permission 0700
        create new settings file in $new_location/
        if creation of settings file failed:
            log warning

The specifics about `<conf-name>` are out of scope of this issue (right now we use `dosbox-staging-git.conf` - it will stay this way for all our development snapshots and as default, but for stable releases it will change. Also, @sharkwouter - I recomend testing if [SDL_GetPrefPath](https://wiki.libsdl.org/SDL_GetPrefPath) fits our requirements.

All other related behaviour (like e.g. skipping primary .conf file when `-conf` parameter is passed, default location for searching for gus patches or GL shaders, etc) - should stay as is implemented right now.

**In scope** of this issue is updating all documentation (README/manual, man page, configuration help strings).

**Out of scope** are:
- XDG user directories, that should be respected but currently aren't (like `$XDG_PICTURES_DIR` as default location for screenshots, instead of `~/.dosbox/`.
- Moving search dir for GUS patches and GL shaders to `$XDG_DATA_HOME`
- Parallel changes on other OSes to make sure native conventions are being followed

We will address those in a subsequent issue/PR (not sure if it qualifies for 0.75.0 release).

**Comment by es20490446e** [comment]
About the future of the project [here](https://github.com/dreamer/dosbox-staging/issues/217).

**Comment by sharkwouter** [comment]
Honestly, because of my very limited knowledge of C++, I've not really found the motivation to work on this issue. Perhaps it would be better to reassign it to someone else.

**Comment by dreamer** [comment]
@sharkwouter that's fine, I'm taking it over. Thanks for the commits you landed in the repo so far and for your packaging work, we're closing in on release, so having AUR package is very helpful :)

**Comment by dreamer** [comment]
@es20490446e @sharkwouter @i30817 The test implementation is ready on branch `po/xdg-config-1`; waiting for your feedback if it works ok :)

**Comment by es20490446e** [comment]
> git clone --single-branch --branch po/xdg-config-1 git@github.com:dosbox-staging/dosbox-staging.git; cd dosbox-staging; ./autogen.sh; ./configure; make

**Comment by es20490446e** [comment]
Right now we have:
~/.config/dosbox/dosbox-staging-${version}.conf

What if the file simply was:
~/.config/dosbox-staging

While having the version variable inside it? That would prevent accumulating version files through time 🤔

**Comment by i30817** [comment]
That's against the standard. All program files must be on a single directory. The version is only on the conf files and that is dosbox upstream policy.

In fact there is code that purposefully separates 'svn' config files from 'release' config files... for some reason.

**Comment by es20490446e** [comment]
Just ideas, in case you wanted to finely knit:

It could be:
~/.config/dosbox-staging/${version}

Where "version" is either "stable" or "development". So you can test both versions simultaneously without them conflicting, but not having more than those two.

**Comment by i30817** [comment]
Many people (unfortunately including the dosbox libretro core people) don't understand how dosbox conf files are 'supposed' to be used. Namely, they have support for onion configuration, and the idea is to keep the master 'conf file' in the main dir with all the defaults or options that don't matter if they're on but not used (like mt32 or ULTRASND dir), and per game configs just change things the game needs, such as autoexec.

Then you run dosbox with `dosbox -conf -userconf GAMECONF.conf`. GAMECONF.conf replaces options from the main conf and thus is 'minimal'.

**Comment by dreamer** [comment]
There will be no versioned configs for dosbox-staging in the future - we discussed it some time ago with @kcgen and decided neither of us likes these versioned files. I know no other software, that generates a new configuration file for each release, it's a bizarre solution.

It'll be:

- `~/.config/dosbox/dosbox-staging.conf` for stable releases
- `~/.config/dosbox/dosbox-staging-git.conf` for all development branches (where we'll experiment with new options, test, and change stuff)

**Comment by es20490446e** [comment]
Fair enough! 👍

**Comment by kcgen** [comment]
Yeah @i30817, your onion description is perfect. We should try to teach or convey this; maybe via the wiki. 

Also, there are two halves to a typical dosbox configuration: 
- what the game needs in terms of emulated hardware. For example, Space Quest 1 wants EGA and aspect correction, low cycle-count, Tandy sound, and doesn't need XMS or EMS memory nor sound blaster or GUS. These emulation-side settings are 1:1 for each game, and should be defined minimally without [sdl] settings.

- Then you've got your user interface settings; or how you want to use DOSBox: fullscreen, mouse locking on start up, rendered with pixel perfect. These are typically defined in only one place, and might be unique per machine.  For example, my Rpi settings here differ significantly from my dev desktop.

**Comment by dreamer** [comment]
#292 just got merged, this change is now in master, as of 64772850b51fe2c3fc4aa4e33bd8f86ba593864a

Thanks for feedback, implementation help, and testing everyone :)
